modules/training.py

```diff
@@ -0,0 +1,236 @@
+import json
+import sys
+import threading
+import time
+from pathlib import Path
+
+import gradio as gr
+import torch
+import transformers
+from datasets import load_dataset
+from peft import (LoraConfig, get_peft_model, get_peft_model_state_dict,
+                  prepare_model_for_int8_training)
+
+from modules import shared, ui
+
+WANT_INTERRUPT = False
+CURRENT_STEPS = 0
+MAX_STEPS = 0
+CURRENT_GRADIENT_ACCUM = 1
+
+def get_json_dataset(path: str):
+    def get_set():
+        return ['None'] + sorted(set(map(lambda x : '.'.join(str(x.name).split('.')[:-1]), Path(path).glob('*.json'))), key=str.lower)
+    return get_set
+
+def create_train_interface():
+    with gr.Tab('Train LoRA', elem_id='lora-train-tab'):
+        lora_name = gr.Textbox(label="Name", info="The name of your new LoRA file")
+        with gr.Row():
+            # TODO: Implement multi-device support.
+            micro_batch_size = gr.Slider(label='Micro Batch Size', value=4, minimum=1, maximum=128, step=1, info='Per-device batch size (NOTE: multiple devices not yet implemented). Increasing this will increase VRAM usage.')
+            batch_size = gr.Slider(label='Batch Size', value=128, minimum=1, maximum=1024, step=4, info='Global batch size. The two batch sizes together determine gradient accumulation (gradientAccum = batch / microBatch). Higher gradient accum values lead to better quality training.')
+
+        with gr.Row():
+            epochs = gr.Number(label='Epochs', value=1, info='Number of times every entry in the dataset should be fed into training. So 1 means feed each item in once, 5 means feed it in five times, etc.')
```

Comment:
Maybe set the `epochs` default to `3` like in https://github.com/tloen/alpaca-lora/blob/main/finetune.py?

Reply:
The ideal default for that is... a complex question. That file is explicitly for tloen's run of Alpaca and so has essentially his particular choices for a specific usage as the defaults. I changed a few to what I think are more likely to be closer to baselines for the average user.
That said, the primary justification for my choice of `epochs=1` was that the average user would likely want to minimize downtime between testing runs, and that is now much easier thanks to the clean `Interrupt` button being available, so 3 might be a good default with that available.

Reply:
>I changed a few to what I think are more likely to be closer to baselines for the average user.

What other default is different? From what I have seen, everything else seems to be the same now.

Reply:
I'm... losing my mind apparently. I could swear I had set the `rank` value higher, but apparently I didn't.

Reply:
I also thought `8` sounded low, but clearly it was enough to instruction-tune llama.

Reply:
The low value honestly probably helps with Alpaca - the actual data in their training set is *really bad* so they it's for the best that they don't train it in, and keep only the format/behavioral understanding.

The sad thing is Meta have an official Instruct-tuned model that looks to be really good, but they never released it :(
(It's discussed with examples at the bottom of the LLaMA arxiv paper)

Reply:
Yes, I saw that too. Everyone is afraid of releasing instruction-tuned models because they are too powerful.

Reply:
I'm more convinced now that rank needs to default higher, as I repeatedly tested and thought my tests weren't working because I forgot to bump rank up. The default 8 retains basically nothing in terms of actual knowledge of terms put in.